src/script/standard.cpp

```diff
@@ -120,37 +185,51 @@ bool Solver(const CScript& scriptPubKey, txnouttype& typeRet, std::vector<std::v
     }
 
     std::vector<unsigned char> data;
-    if (MatchPayToPubkey(scriptPubKey, data)) {
-        typeRet = TX_PUBKEY;
+    if (MatchPayToPubkey(scriptPubKey, data, typeRet)) {
         vSolutionsRet.push_back(std::move(data));
         return true;
     }
 
-    if (MatchPayToPubkeyHash(scriptPubKey, data)) {
-        typeRet = TX_PUBKEYHASH;
+    if (MatchPayToPubkeyHash(scriptPubKey, data, typeRet)) {
         vSolutionsRet.push_back(std::move(data));
         return true;
     }
 
     unsigned int required;
     std::vector<std::vector<unsigned char>> keys;
-    if (MatchMultisig(scriptPubKey, required, keys)) {
-        typeRet = TX_MULTISIG;
+    if (MatchMultisig(scriptPubKey, required, keys, typeRet)) {
         vSolutionsRet.push_back({static_cast<unsigned char>(required)}); // safe as required is in range 1..16
         vSolutionsRet.insert(vSolutionsRet.end(), keys.begin(), keys.end());
         vSolutionsRet.push_back({static_cast<unsigned char>(keys.size())}); // safe as size is in range 1..16
         return true;
     }
 
+    if ((typeRet == TX_PUBKEY ||
+        typeRet == TX_PUBKEYHASH ||
+        typeRet == TX_MULTISIG ||
+        typeRet == TX_SCRIPTHASH) &&
+        vSolutionsRet.empty())
+        {
+            typeRet = TX_NONSTANDARD;
+            return false;
+        }
+
```

Comment:
not need this check. It is the responsibility of the user who created the script to make sure the script is correct.